src/libCli/ConnectionManager.cpp

```diff
@@ -0,0 +1,215 @@
+
+// Copyright 2019 IBM Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "libCli/ConnectionManager.hpp"
+#include <gRPC_utils/proto_reflection_descriptor_database.h>
+
+#include "utils/gwhisperUtils.hpp"
+#include "libLocalDescriptorCache/DescDbProxy.hpp"
+
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/security/credentials.h>
+#include <libArgParse/ArgParse.hpp>
+
+namespace cli
+{
+    ConnectionManager &ConnectionManager::getInstance() //Warum brauche ich static nicht mehr hier davor?
+    {
+        static ConnectionManager connectionManager;
+        return connectionManager;
+    }
+
+    std::shared_ptr<grpc::Channel> ConnectionManager::getChannel(std::string f_serverAddress, ArgParse::ParsedElement &f_parseTree)
+    {
+        if (!findChannelByAddress(f_serverAddress))
+        {
+            registerConnection(f_serverAddress, f_parseTree);
+        }
+        return m_connections[f_serverAddress].channel;
+    }
+
+    std::shared_ptr<DescDbProxy> ConnectionManager::getDescDb(std::string f_serverAddress, ArgParse::ParsedElement &f_parseTree)
+    {
+        if (!findDescDbByAddress(f_serverAddress))
+        {
+            if (m_connections[f_serverAddress].channel)
+            {
+                ensureDescDbProxyAndDescPoolIsAvailable(f_serverAddress,f_parseTree);
+            }
+            else
+            {
+                registerConnection(f_serverAddress, f_parseTree);
+            }
+        }
+        return m_connections[f_serverAddress].descDbProxy;
+    }
+    
+    std::shared_ptr<grpc::protobuf::DescriptorPool> ConnectionManager::getDescPool(std::string f_serverAddress, ArgParse::ParsedElement &f_parseTree)
+    {
+        if (!findDescPoolByAddress(f_serverAddress))
+        {
+            ensureDescDbProxyAndDescPoolIsAvailable(f_serverAddress, f_parseTree);
+        }
+        return m_connections[f_serverAddress].descPool;
+    }
+
+    void ConnectionManager::ensureDescDbProxyAndDescPoolIsAvailable(std::string &f_serverAddress, ArgParse::ParsedElement &f_parseTree)
+    {
+        if (m_connections[f_serverAddress].channel)
```

Comment:
how can this happen, that we have a channel already registered, but no descdb proxy?